### PR TITLE
add banner support to system api

### PR DESCRIPTION
**Pull request by dathelen**
TAC had a requirement for an internal tool to change the banner on lab switches programmatically, so I thought this would be a nice add to the pyeapi library

### CODE REVIEW

**Comment by phil-arista** [comment]
This is awesome and looks really good!  Just some comments:

- don't commit dut changes in test/fixtures/dut.conf
- Add more to your unit tests.  You can add a banner in the test/fixtures flat config files and then process those login banners or motd without connecting to a dut. 

We'll have the next internal review on 1/4 to collect any remaining comments.  Thanks again!!

**Comment by dathelen** [comment]
Sounds good Phil.  I forgot to revert my dut config before I committed.  I
will get that backed out and write some unit tests.  Thanks for looking
into it!

-

File: pyeapi/client.py [hunk 1]
```diff
@@ -522,6 +522,43 @@ def config(self, commands):
...
+               changes.   The function will preprend the required commands
+               to put the session in config mode
+ 
+          Returns:
+              The config method will return a list of dictionaries with 
+              the ouptut from each command
+        """ 
+        command_list = [] 
+        command_list.append('configure terminal')
+        if isinstance(commands, list):
```
Comment (devrobo) [inline pyeapi/client.py]: I think you can make the change to the config method by checking for the value key for each command. If the value key is defined then you would add the input key when building the command list. Note that you can leverage the make_iterable call in the config method.

**Comment by devrobo** [comment]
If you have any questions about the change requested we would be happy to setup a call to discuss further.

File: pyeapi/client.py [hunk 2]
```diff
@@ -522,6 +538,43 @@ def config(self, commands):
         response.pop(0)
 
         return response
+    def config_with_input(self, commands):
```
Comment (devrobo) [inline pyeapi/client.py]: I don't think you need this method anymore.

File: test/fixtures/dut.conf [hunk 3]
```diff
@@ -3,3 +3,4 @@ host: 192.168.1.16
 username: eapi
 password: password
 transport: http
+
```
Comment (devrobo) [inline test/fixtures/dut.conf]: Unnecessary empty line, there shouldn't be any changes to the dut.conf.

Comment (dathelen) [inline test/fixtures/dut.conf]: Thanks John.  Sorry about the oversights here.

**Comment by grybak-arista** [comment]
Just a few more changes and I think we'll be ready to merge.

We have decided that there shouldn't be any changes to pyeapi/client.py for this addition. Please back out those changes and handle the construction of the {cmd, input} dict in the set_banner method, and pass it as a single element list to self.configure. So if default or disable are True or value is None, use command_builder, otherwise create the command input:
     command_input={'cmd': command_string, 'input': value}
     self.configure([command_input])

Also, please update the regex in _parse_banners to '^banner\s+(login|motd)\s?$\n(.*?)$\nEOF$\n'
This is a more robust expression that allows EOF to be part of the banner itself, as in the following examples:

banner login
Login Banner
Second Line
123456
  EOF     <--- Space before EOF
EOF
!
banner motd
MOTD Banner
Second Line
EOF      <--- Space after EOF
*\v1?
EOF
!

Please note that you will need to update the test cases to set the initial banner values using the dictionary format: dut.config([{'cmd': 'banner login', 'input': banner_value}])

Let me know if something doesn't make sense.
Thanks!

**Comment by dathelen** [comment]
I have integrated all of the changes you have asked for.  It does appear that new _parse_banners regex you have suggested is causing some of the system tests to fail.  I am just looking into whether its just the way I have written test or an issue with the regular expression.

**Comment by dathelen** [comment]
I apologize in advanced about the long windedness of this response.

It appears that setting a banner via the eAPI interface on EOS will not put the EOF on a newline implicitly like it forces you do explicitly if you set the banner via the CLI.  This seems a like bug (albeit nitpicky) in EOS, but it is the root cause of what I am seeing in the system tests I wrote.  Here is more detail.

I took pyeapi out of it just to make sure something wasnt going wrong with the methods I wrote.  If you set a banner without a new line like so:

```
>>> from jsonrpclib import Server
>>> eapi = Server("https://eapi:eapi@localhost/command-api")
>>> eapi.runCmds(1, ["enable", "configure terminal", {"cmd":"banner motd", "input":"foo"}])
[{}, {}, {}]
```

Then if you look at the running config on the switch using the CLI you see the following:

```
banner motd
fooEOF
```

If the eAPI call adds an explicit newline at the end of the input value like:

```
>>> eapi.runCmds(1, ["enable", "configure terminal", {"cmd":"banner motd", "input":"foo\n"}])
[{}, {}, {}]
```
The banner looks as expected:
```
banner motd
foo
EOF
```

FWIW, even when the EOF is not on its own line the show commands on the CLI still work as expected.

Since, in the system tests I wrote I was just calling randon_string(), it wasn't adding a newline explicitly. Which was causing the new regex to fail because EOF wasn't on its own line.

I think we have a few options:

1) I can check for a trailing new line in the set_banner method, and if I don't find one add one (that does add an extra newline to show banner output, FWIW).
2) file a bug with EOS and implement the new regex and change my system tests to add a new line explicitly

Let me know which one fits best here for pyeapi.

**Comment by devrobo** [comment]
I hit this problem when doing the rbeapi version. Choice 1 works and shouldn't be a problem. If you entered the banner from the CLI you would have to enter a newline before typing EOF so the end result should be the same unless I am missing something.

**Comment by dathelen** [comment]
Thanks John.  I should have that done soon.

**Comment by rmadsen** [comment]
I vote that we open a BUG on EOS—we should make eAPI as intuitive as possible.

**Comment by rmadsen** [comment]
Actually, we've already fixed this behavior in EOS starting in 4.15.2F: multiline CLI inputs will now add a trailing `\n` if none is present.

**Comment by dathelen** [comment]
thanks Ryan

**Pull request merged**